File: backend/features/patrols/routes/patrolRoutes.js

const multer = require("multer");
const upload = multer({ storage: multer.memoryStorage() });

const express = require("express");
const router  = express.Router();
const {
  getMyPatrols,
  getAfterPatrolReports,
  getMyAfterPatrolReports,
  getPatrolStats,
  getActivePatrollers,
  getAvailablePatrollers,
  getAvailableMobileUnits,
  getMobileUnits,
  createMobileUnit,
  submitAfterPatrolReport,
  updateMobileUnit,
deleteAfterPatrolReport,
  deleteMobileUnit,
  getPatrols,
  createPatrol,
  updatePatrol,
  updatePatrollersForDate,
  deletePatrol,
  updateRouteNotes,
  updateRouteTask,
  addRouteTask,
  removeRouteTask,
  updateOfficerLocation,
   uploadAfterPatrolPhotos,
  deleteAfterPatrolPhoto,
} = require("../controllers/patrolController");

const { authenticate } = require("../../../shared/middleware/tokenMiddleware");

const { exportPatrolList, exportPatrolDetail } = require("../controllers/ExportPatrolController");

// Stats & listings
router.get("/stats",               authenticate, getPatrolStats);
router.get("/active",              authenticate, getActivePatrollers);
router.get("/available-patrollers",authenticate, getAvailablePatrollers);
router.get("/available-mobile-units", getAvailableMobileUnits);

// Mobile units
router.get   ("/mobile-units",     authenticate, getMobileUnits);
router.post  ("/mobile-units",     authenticate, createMobileUnit);
router.put   ("/mobile-units/:id", authenticate, updateMobileUnit);
router.delete("/mobile-units/:id", authenticate, deleteMobileUnit);

// Patrols
router.get("/my-patrols", authenticate, getMyPatrols);
router.get   ("/patrols",     authenticate, getPatrols);
router.post  ("/patrols",     authenticate, createPatrol);
router.put   ("/patrols/:id", authenticate, updatePatrol);
router.delete("/patrols/:id", authenticate, deletePatrol);


// ── Patrollers per date (new) ──────────────────────────────
// PATCH /patrol/patrols/:id/patrollers/:date
// Replace all patrollers for a specific patrol date
router.patch("/patrols/:id/patrollers/:date", authenticate, updatePatrollersForDate);

// Routes / tasks
router.patch ("/routes/:routeId/notes", authenticate, updateRouteNotes);
router.patch ("/routes/:routeId/task",  authenticate, updateRouteTask);
router.post  ("/routes/add",            authenticate, addRouteTask);
router.delete("/routes/:routeId",       authenticate, removeRouteTask);

//
router.post("/export/list",   authenticate, exportPatrolList);
router.post("/export/detail", authenticate, exportPatrolDetail);

//
 router.post("/patrols/:id/after-report", authenticate, submitAfterPatrolReport);
router.get( "/patrols/:id/after-reports", authenticate, getAfterPatrolReports);
router.get( "/patrols/:id/after-reports/mine", authenticate, getMyAfterPatrolReports);
router.delete("/after-reports/:reportId", authenticate, deleteAfterPatrolReport)

//
router.post("/location", authenticate, updateOfficerLocation);
module.exports = router;

//picture
router.post(
  "/after-reports/:reportId/photos",
  authenticate,
  upload.array("photos", 10),
  uploadAfterPatrolPhotos
);
router.delete(
  "/after-reports/:reportId/photos",
  authenticate,
  deleteAfterPatrolPhoto
);

File: backend/features/patrols/patrol_routes.py
# ================================================================================
# FILE: backend/features/patrols/patrol_routes.py
# ================================================================================

from flask import Blueprint
from shared.middleware.token_middleware import authenticate
from features.patrols.patrol_controller import (
    get_patrol_stats,
    get_active_patrollers,
    get_available_patrollers,
    get_available_mobile_units,
    get_mobile_units,
    create_mobile_unit,
    update_mobile_unit,
    delete_mobile_unit,
    get_my_patrols,
    get_patrols,
    create_patrol,
    update_patrol,
    delete_patrol,
    update_patrollers_for_date,
    update_route_notes,
    update_route_task,
    add_route_task,
    remove_route_task,
    submit_after_patrol_report,
    get_after_patrol_reports,
    get_my_after_patrol_reports,
    delete_after_patrol_report,
    update_officer_location,
    upload_after_patrol_photos,
    delete_after_patrol_photo,
)
from features.patrols.export_patrol_controller import (
    export_patrol_list,
    export_patrol_detail,
)

patrol_bp = Blueprint("patrol", __name__)

# ── Stats & listings ──────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/stats",                  view_func=authenticate(get_patrol_stats),         methods=["GET"])
patrol_bp.add_url_rule("/active",                 view_func=authenticate(get_active_patrollers),    methods=["GET"])
patrol_bp.add_url_rule("/available-patrollers",   view_func=authenticate(get_available_patrollers), methods=["GET"])
patrol_bp.add_url_rule("/available-mobile-units", view_func=get_available_mobile_units,             methods=["GET"])

# ── Mobile units ──────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/mobile-units",      view_func=authenticate(get_mobile_units),   methods=["GET"])
patrol_bp.add_url_rule("/mobile-units",      view_func=authenticate(create_mobile_unit), methods=["POST"])
patrol_bp.add_url_rule("/mobile-units/<id>", view_func=authenticate(update_mobile_unit), methods=["PUT"])
patrol_bp.add_url_rule("/mobile-units/<id>", view_func=authenticate(delete_mobile_unit), methods=["DELETE"])

# ── Patrols ───────────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/my-patrols",   view_func=authenticate(get_my_patrols),  methods=["GET"])
patrol_bp.add_url_rule("/patrols",      view_func=authenticate(get_patrols),     methods=["GET"])
patrol_bp.add_url_rule("/patrols",      view_func=authenticate(create_patrol),   methods=["POST"])
patrol_bp.add_url_rule("/patrols/<id>", view_func=authenticate(update_patrol),   methods=["PUT"])
patrol_bp.add_url_rule("/patrols/<id>", view_func=authenticate(delete_patrol),   methods=["DELETE"])

# ── Patrollers per date ───────────────────────────────────────────────────────
patrol_bp.add_url_rule("/patrols/<id>/patrollers/<date>", view_func=authenticate(update_patrollers_for_date), methods=["PATCH"])

# ── Routes / tasks ────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/routes/<route_id>/notes", view_func=authenticate(update_route_notes), methods=["PATCH"])
patrol_bp.add_url_rule("/routes/<route_id>/task",  view_func=authenticate(update_route_task),  methods=["PATCH"])
patrol_bp.add_url_rule("/routes/add",              view_func=authenticate(add_route_task),     methods=["POST"])
patrol_bp.add_url_rule("/routes/<route_id>",       view_func=authenticate(remove_route_task),  methods=["DELETE"])

# ── Export ────────────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/export/list",   view_func=authenticate(export_patrol_list),   methods=["POST"])
patrol_bp.add_url_rule("/export/detail", view_func=authenticate(export_patrol_detail), methods=["POST"])

# ── After patrol reports ──────────────────────────────────────────────────────
patrol_bp.add_url_rule("/patrols/<id>/after-report",       view_func=authenticate(submit_after_patrol_report),  methods=["POST"])
patrol_bp.add_url_rule("/patrols/<id>/after-reports",      view_func=authenticate(get_after_patrol_reports),    methods=["GET"])
patrol_bp.add_url_rule("/patrols/<id>/after-reports/mine", view_func=authenticate(get_my_after_patrol_reports), methods=["GET"])
patrol_bp.add_url_rule("/after-reports/<report_id>",       view_func=authenticate(delete_after_patrol_report),  methods=["DELETE"])

# ── Photos ────────────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/after-reports/<report_id>/photos", view_func=authenticate(upload_after_patrol_photos), methods=["POST"])
patrol_bp.add_url_rule("/after-reports/<report_id>/photos", view_func=authenticate(delete_after_patrol_photo),  methods=["DELETE"])

# ── Location ──────────────────────────────────────────────────────────────────
patrol_bp.add_url_rule("/location", view_func=authenticate(update_officer_location), methods=["POST"])